torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -653,8 +653,10 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
          << " milliseconds and timed out.";
       const auto exceptionMsg = createExceptionResponse(
           Message({}, {}, MessageType::EXCEPTION), ss.str());
-      timedOutFuture.future_->setError(std::string(
-          exceptionMsg.payload().begin(), exceptionMsg.payload().end()));
+      if (!timedOutFuture.future_->hasError()) {
```

Comment:
@osalpekar Why we need this? 
We shouldn't expect these future pulled out from the map has been used by other threads, right?

Reply:
I added this since our method of testing involved sending failed RPCs (by calling send and immediately marking the returned future with an error). Checking whether the future already has an error here avoided double-marking the future with an error (which I believe is illegal). There's a diff out to change the failing-sends implementation (currently in ProcessGroupAgentTest.cpp) by just returning a future marked with an error (and never actually calling `ProcessGroupAgent::send`), and with that change, this line could probably be safely removed.

Reply:
@xush6528 Agree that this is a race condition, since there's a time gap between evaluating and setting.

I ran into this in an unrelated setting (my rpc server async change), and I think it may now be unnecessary for the tests with [pending] D20317773 fix.